File: src/lib/tmdb.ts
import { db } from "@/lib/db";
import { settings } from "@/lib/db/schema";
import { eq } from "drizzle-orm";

const TMDB_BASE = "https://api.themoviedb.org/3";

// Simple rate limiter: max 40 requests per 10 seconds (TMDB limit)
let requestTimestamps: number[] = [];
const RATE_LIMIT = 35; // stay slightly under
const RATE_WINDOW = 10000;

async function rateLimitedFetch(url: string, init?: RequestInit): Promise<Response> {
  const now = Date.now();
  requestTimestamps = requestTimestamps.filter((t) => now - t < RATE_WINDOW);

  if (requestTimestamps.length >= RATE_LIMIT) {
    const waitTime = RATE_WINDOW - (now - requestTimestamps[0]);
    await new Promise((resolve) => setTimeout(resolve, waitTime));
  }

  requestTimestamps.push(Date.now());
  return fetch(url, init);
}

function getApiKey(): string {
  const setting = db
    .select()
    .from(settings)
    .where(eq(settings.key, "tmdb_api_key"))
    .get();
  return setting?.value || "";
}

export interface TmdbSearchResult {
  id: number;
  title?: string;
  name?: string;
  release_date?: string;
  first_air_date?: string;
  poster_path: string | null;
  overview: string;
  popularity: number;
  media_type?: string;
  vote_average: number;
}

export interface TmdbSearchResponse {
  results: TmdbSearchResult[];
  total_results: number;
}

export async function searchMulti(
  query: string,
  year?: number
): Promise<TmdbSearchResult[]> {
  const apiKey = getApiKey();
  if (!apiKey) throw new Error("TMDB API key not configured");

  const params = new URLSearchParams({
    api_key: apiKey,
    query,
    include_adult: "false",
  });
  if (year) params.set("year", String(year));

  const res = await rateLimitedFetch(`${TMDB_BASE}/search/multi?${params}`);
  if (!res.ok) throw new Error(`TMDB API error: ${res.status}`);

  const data: TmdbSearchResponse = await res.json();
  // Filter to only movies and TV shows
  return data.results.filter(
    (r) => r.media_type === "movie" || r.media_type === "tv"
  );
}

export async function searchMovies(
  query: string,
  year?: number
): Promise<TmdbSearchResult[]> {
  const apiKey = getApiKey();
  if (!apiKey) throw new Error("TMDB API key not configured");

  const params = new URLSearchParams({
    api_key: apiKey,
    query,
    include_adult: "false",
  });
  if (year) params.set("year", String(year));

  const res = await rateLimitedFetch(`${TMDB_BASE}/search/movie?${params}`);
  if (!res.ok) throw new Error(`TMDB API error: ${res.status}`);

  const data: TmdbSearchResponse = await res.json();
  return data.results.map((r) => ({ ...r, media_type: "movie" }));
}

export async function searchTV(
  query: string,
  year?: number
): Promise<TmdbSearchResult[]> {
  const apiKey = getApiKey();
  if (!apiKey) throw new Error("TMDB API key not configured");

  const params = new URLSearchParams({
    api_key: apiKey,
    query,
    include_adult: "false",
  });
  if (year) params.set("first_air_date_year", String(year));

  const res = await rateLimitedFetch(`${TMDB_BASE}/search/tv?${params}`);
  if (!res.ok) throw new Error(`TMDB API error: ${res.status}`);

  const data: TmdbSearchResponse = await res.json();
  return data.results.map((r) => ({ ...r, media_type: "tv" }));
}

export interface TmdbSeason {
  id: number;
  name: string;
  season_number: number;
  episode_count: number;
  air_date: string | null;
  overview: string;
  poster_path: string | null;
}

export interface TmdbSeasonDetail {
  id: number;
  name: string;
  season_number: number;
  episodes: TmdbEpisode[];
}

export async function getShowSeasons(tvId: number): Promise<TmdbSeason[]> {
  const apiKey = getApiKey();
  if (!apiKey) throw new Error("TMDB API key not configured");

  const res = await rateLimitedFetch(
    `${TMDB_BASE}/tv/${tvId}?api_key=${apiKey}`
  );
  if (!res.ok) throw new Error(`TMDB API error: ${res.status}`);

  const data = await res.json();
  return data.seasons || [];
}

export async function getSeason(
  tvId: number,
  seasonNumber: number
): Promise<TmdbSeasonDetail | null> {
  const apiKey = getApiKey();
  if (!apiKey) throw new Error("TMDB API key not configured");

  const res = await rateLimitedFetch(
    `${TMDB_BASE}/tv/${tvId}/season/${seasonNumber}?api_key=${apiKey}`
  );
  if (!res.ok) return null;

  return res.json();
}

export interface TmdbEpisode {
  id: number;
  name: string;
  episode_number: number;
  season_number: number;
  overview: string;
  still_path: string | null;
}

export async function getEpisode(
  tvId: number,
  season: number,
  episode: number
): Promise<TmdbEpisode | null> {
  const apiKey = getApiKey();
  if (!apiKey) return null;

  const res = await rateLimitedFetch(
    `${TMDB_BASE}/tv/${tvId}/season/${season}/episode/${episode}?api_key=${apiKey}`
  );
  if (!res.ok) return null;

  return res.json();
}

File: src-tauri/src/tmdb.rs
use anyhow::{anyhow, Result};
use once_cell::sync::Lazy;
use reqwest::Client;
use serde::{Deserialize, Serialize};
use std::sync::Mutex;
use std::time::Instant;
use tokio::time::sleep;
use tracing::debug;

const TMDB_BASE: &str = "https://api.themoviedb.org/3";
const RATE_LIMIT: usize = 35;
const RATE_WINDOW_MS: u128 = 10_000;

static HTTP_CLIENT: Lazy<Client> = Lazy::new(|| Client::new());
static REQUEST_TIMESTAMPS: Lazy<Mutex<Vec<Instant>>> = Lazy::new(|| Mutex::new(Vec::new()));

async fn rate_limited_get(url: &str) -> Result<reqwest::Response> {
    // Check rate limit and compute wait time (drop guard before any .await)
    let wait_ms = {
        let mut timestamps = REQUEST_TIMESTAMPS.lock().unwrap();
        let now = Instant::now();
        timestamps.retain(|t| now.duration_since(*t).as_millis() < RATE_WINDOW_MS);

        if timestamps.len() >= RATE_LIMIT {
            let wait = RATE_WINDOW_MS - now.duration_since(timestamps[0]).as_millis();
            Some(wait)
        } else {
            timestamps.push(now);
            None
        }
    }; // MutexGuard dropped here

    if let Some(wait) = wait_ms {
        debug!("Rate limit reached, waiting {}ms", wait);
        sleep(std::time::Duration::from_millis(wait as u64)).await;
        let mut timestamps = REQUEST_TIMESTAMPS.lock().unwrap();
        timestamps.push(Instant::now());
    }

    let res = HTTP_CLIENT.get(url).send().await?;
    Ok(res)
}

fn get_api_key() -> Result<String> {
    crate::db::get_setting("tmdb_api_key")
        .filter(|k| !k.is_empty())
        .ok_or_else(|| anyhow!("TMDB API key not configured"))
}

// ── Response types ──────────────────────────────────────

#[derive(Debug, Clone, Deserialize)]
pub struct TmdbSearchResult {
    pub id: i64,
    pub title: Option<String>,
    pub name: Option<String>,
    pub release_date: Option<String>,
    pub first_air_date: Option<String>,
    pub poster_path: Option<String>,
    pub overview: Option<String>,
    pub popularity: Option<f64>,
    pub media_type: Option<String>,
    pub vote_average: Option<f64>,
}

#[derive(Debug, Deserialize)]
struct SearchResponse {
    results: Vec<TmdbSearchResult>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct TmdbSeason {
    pub id: i64,
    pub name: String,
    pub season_number: i64,
    pub episode_count: i64,
    pub air_date: Option<String>,
    pub overview: Option<String>,
    pub poster_path: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct TmdbSeasonDetail {
    pub id: i64,
    pub name: String,
    pub season_number: i64,
    pub episodes: Vec<TmdbEpisode>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct TmdbEpisode {
    pub id: i64,
    pub name: String,
    pub episode_number: i64,
    pub season_number: i64,
    pub overview: Option<String>,
    pub still_path: Option<String>,
}

#[derive(Debug, Deserialize)]
struct TvDetailResponse {
    seasons: Option<Vec<TmdbSeason>>,
}

// ── Public API ──────────────────────────────────────────

pub async fn search_multi(query: &str, year: Option<i64>) -> Result<Vec<TmdbSearchResult>> {
    let api_key = get_api_key()?;
    let mut url = format!(
        "{}/search/multi?api_key={}&query={}&include_adult=false",
        TMDB_BASE,
        api_key,
        urlencoding::encode(query)
    );
    if let Some(y) = year {
        url.push_str(&format!("&year={}", y));
    }

    let res = rate_limited_get(&url).await?;
    if !res.status().is_success() {
        return Err(anyhow!("TMDB API error: {}", res.status()));
    }

    let data: SearchResponse = res.json().await?;
    Ok(data
        .results
        .into_iter()
        .filter(|r| {
            matches!(
                r.media_type.as_deref(),
                Some("movie") | Some("tv")
            )
        })
        .collect())
}

pub async fn search_movies(query: &str, year: Option<i64>) -> Result<Vec<TmdbSearchResult>> {
    let api_key = get_api_key()?;
    let mut url = format!(
        "{}/search/movie?api_key={}&query={}&include_adult=false",
        TMDB_BASE,
        api_key,
        urlencoding::encode(query)
    );
    if let Some(y) = year {
        url.push_str(&format!("&year={}", y));
    }

    let res = rate_limited_get(&url).await?;
    if !res.status().is_success() {
        return Err(anyhow!("TMDB API error: {}", res.status()));
    }

    let data: SearchResponse = res.json().await?;
    Ok(data
        .results
        .into_iter()
        .map(|mut r| {
            r.media_type = Some("movie".to_string());
            r
        })
        .collect())
}

pub async fn search_tv(query: &str, year: Option<i64>) -> Result<Vec<TmdbSearchResult>> {
    let api_key = get_api_key()?;
    let mut url = format!(
        "{}/search/tv?api_key={}&query={}&include_adult=false",
        TMDB_BASE,
        api_key,
        urlencoding::encode(query)
    );
    if let Some(y) = year {
        url.push_str(&format!("&first_air_date_year={}", y));
    }

    let res = rate_limited_get(&url).await?;
    if !res.status().is_success() {
        return Err(anyhow!("TMDB API error: {}", res.status()));
    }

    let data: SearchResponse = res.json().await?;
    Ok(data
        .results
        .into_iter()
        .map(|mut r| {
            r.media_type = Some("tv".to_string());
            r
        })
        .collect())
}

pub async fn get_show_seasons(tv_id: i64) -> Result<Vec<TmdbSeason>> {
    let api_key = get_api_key()?;
    let url = format!("{}/tv/{}?api_key={}", TMDB_BASE, tv_id, api_key);

    let res = rate_limited_get(&url).await?;
    if !res.status().is_success() {
        return Err(anyhow!("TMDB API error: {}", res.status()));
    }

    let data: TvDetailResponse = res.json().await?;
    Ok(data.seasons.unwrap_or_default())
}

pub async fn get_season(tv_id: i64, season_number: i64) -> Result<Option<TmdbSeasonDetail>> {
    let api_key = get_api_key()?;
    let url = format!(
        "{}/tv/{}/season/{}?api_key={}",
        TMDB_BASE, tv_id, season_number, api_key
    );

    let res = rate_limited_get(&url).await?;
    if !res.status().is_success() {
        return Ok(None);
    }

    let data: TmdbSeasonDetail = res.json().await?;
    Ok(Some(data))
}

pub async fn get_episode(
    tv_id: i64,
    season: i64,
    episode: i64,
) -> Result<Option<TmdbEpisode>> {
    let api_key = get_api_key()?;
    let url = format!(
        "{}/tv/{}/season/{}/episode/{}?api_key={}",
        TMDB_BASE, tv_id, season, episode, api_key
    );

    let res = rate_limited_get(&url).await?;
    if !res.status().is_success() {
        return Ok(None);
    }

    let data: TmdbEpisode = res.json().await?;
    Ok(Some(data))
}
